packages/flutter/lib/src/widgets/expansible.dart

```diff
@@ -0,0 +1,377 @@
+///
+/// The `header` property is the header returned by [Expansible.headerBuilder].
+/// The `body` property is the body returned by [Expansible.bodyBuilder] wrapped
+/// in an [Offstage] to hide the body when the [Expansible] is collapsed.
+///
+/// The `animation` property exposes the underlying expanding or collapsing
+/// animation, which has a value of 0 when the [Expansible] is completely
+/// collapsed and 1 when it is completely expanded. This can be used to drive
+/// animations that sync up with the expanding or collapsing animation, such as
+/// rotating an icon.
+///
+/// See also:
+///
+///   * [Expansible.expansibleBuilder], which is of this type.
+typedef ExpansibleBuilder =
+    Widget Function(BuildContext context, Widget header, Widget body, Animation<double> animation);
+
+/// A controller for managing the expansion state of an [Expansible].
+///
+/// This class is a [ChangeNotifier] that notifies its listeners if the value of
+/// [isExpanded] changes.
+///
+/// This controller provides methods to programmatically expand or collapse the
+/// widget, and it allows external components to query the current expansion
+/// state.
+///
+/// The controller's [expand] and [collapse] methods cause the
+/// the [Expansible] to rebuild, so they may not be called from
+/// a build method.
+///
+/// Remember to [dispose] of the [ExpansibleController] when it is no longer
+/// needed. This will ensure we discard any resources used by the object.
+class ExpansibleController extends ChangeNotifier {
+  /// Creates a controller to be used with [Expansible.controller].
+  ExpansibleController();
+
+  _ExpansibleState? _state;
+
+  /// Whether the expansible widget built with this controller is in expanded
+  /// state.
+  ///
+  /// This property doesn't take the animation into account. It reports `true`
+  /// even if the expansion animation is not completed.
+  ///
+  /// See also:
+  ///
+  ///  * [expand], which expands the expansible widget.
+  ///  * [collapse], which collapses the expansible widget.
+  bool get isExpanded => _isExpanded;
+  bool _isExpanded = false;
+  set isExpanded(bool value) {
+    if (_isExpanded == value) {
+      return;
+    }
+    _isExpanded = value;
+    notifyListeners();
+  }
+
+  /// Expands the expansible widget that was built with this controller.
+  ///
+  /// If the widget is already in the expanded state (see [isExpanded]), calling
+  /// this method has no effect.
+  ///
+  /// Calling this method may cause the [Expansible] to rebuild, so it may
+  /// not be called from a build method.
+  ///
+  /// Calling this method will notify registered listeners of this controller
+  /// that the expansion state has changed.
+  ///
+  /// See also:
+  ///
+  ///  * [collapse], which collapses the expansible widget.
+  ///  * [isExpanded] to check whether the expansible widget is expanded.
+  void expand() {
+    assert(_state != null);
+    if (!_isExpanded) {
+      _state!._toggleExpansion();
```

Comment:
This doesn't need to be tied to `_state`, instead the _state should listen to this controller and call _toggleExpansion itself

Reply:
this should just set isExpanded

Reply:
Why do we have both a `set isExpanded` and `expand()`? Aren't they basically the same (and at least should be implemented so)?